File: gui/menu/compute/compute_model.cpp
#include "compute_model.h"


/* NOTE FOR THE DEVELOPERS:
 * The constructor of a Menu Dialog class deriving from QIodeSettings MUST:
 * 2. call setupUi(this),
 * 3. use the Wrapper classes defined in the wrapper_classes.h header file to wrap the input field items,
 * 4. fill the mapFields member (inherited from QIodeSettings) with the wrapped field items,
 * 5. initialize the inherited className member (using typeid(this).name())
 * 6. call the loadSettings() method (inherited from QIodeSettings).
 */


QIodeMenuComputeModel::QIodeMenuComputeModel(const QString& settings_filepath, QWidget* parent, Qt::WindowFlags f) : 
    QIodeSettings(settings_filepath, parent, f)
{
    setupUi(this);

    qEquationsList = new WrapperQTextEdit(label_equations_list->text(), *textEdit_equations_list, OPTIONAL_FIELD);
    mapFields["EquationsList"] = qEquationsList;

    // TODO: if possible, find a way to initialize className inside MixingSettings
    // NOTE FOR DEVELOPPERS: we cannot simply call the line below from the constructor of MixingSettings 
    //                       since in that case this refers to MixingSettings and NOT the derived class
    className = QString::fromStdString(typeid(this).name());
    loadSettings();
}

QIodeMenuComputeModel::~QIodeMenuComputeModel()
{
    delete qEquationsList;
}

void QIodeMenuComputeModel::compute()
{
    try
    {
        std::string equations_list = qEquationsList->extractAndVerify().toStdString();

        Simulation sim;

        sim.model_compile(equations_list);

        this->accept();
    }
    catch (const std::exception& e)
    {
        QMessageBox::critical(this, tr("ERROR"), tr(e.what()));
    }
}

File: gui/menu/compute/compute_model.py
from PySide6.QtCore import Slot
from PySide6.QtWidgets import QWidget, QMessageBox

from settings import MixinSettingsDialog
from .ui_compute_model import Ui_MenuComputeModel

from iode import IodeType, Simulation


class MenuComputeModel(MixinSettingsDialog):
    def __init__(self, parent: QWidget=None):
        super().__init__(parent)
        self.ui = Ui_MenuComputeModel()
        self.ui.setupUi(self)
        self.prepare_settings(self.ui)  

        self.ui.textEdit_equations_list.handle_iode_type(IodeType.EQUATIONS)        
        self.ui.textEdit_equations_list.include_iode_command(False)        
        self.ui.textEdit_equations_list.include_lec_functions(False)

        self.load_settings()

    @Slot()
    def compute(self):
        try:
            equations_list: str = self.ui.textEdit_equations_list.toPlainText().strip()

            simu = Simulation()
            simu.model_compile(equations_list)         

            self.accept()
        except Exception as e:
            QMessageBox.warning(self, "WARNING", str(e))
